indexing-service/src/main/java/org/apache/druid/indexing/common/actions/RetrieveLockedSegmentsAction.java

```diff
@@ -0,0 +1,158 @@
+ */
+
+package org.apache.druid.indexing.common.actions;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.common.collect.ImmutableList;
+import org.apache.druid.indexing.common.task.Task;
+import org.apache.druid.indexing.overlord.Segments;
+import org.apache.druid.metadata.ReplaceTaskLock;
+import org.apache.druid.timeline.DataSegment;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * This TaskAction returns a collection of segments which have data within the specified interval and are marked as
+ * used, and have been created before a REPLACE lock, if any, was acquired.
+ *
+ * The order of segments within the returned collection is unspecified, but each segment is guaranteed to appear in
+ * the collection only once.
+ *
+ * @implNote This action doesn't produce a {@link Set} because it's implemented via {@link
+ * org.apache.druid.indexing.overlord.IndexerMetadataStorageCoordinator#retrieveUsedSegmentsForIntervals} which returns
+ * a collection. Producing a {@link Set} would require an unnecessary copy of segments collection.
+ */
+public class RetrieveLockedSegmentsAction implements TaskAction<Collection<DataSegment>>
+{
+  @JsonIgnore
+  private final String dataSource;
+
+  @JsonIgnore
+  private final Interval interval;
+
+  @JsonIgnore
+  private final Segments visibility;
+
+  @JsonCreator
+  public RetrieveLockedSegmentsAction(
+      @JsonProperty("dataSource") String dataSource,
+      @JsonProperty("interval") Interval interval,
+      // When JSON object is deserialized, this parameter is optional for backward compatibility.
+      // Otherwise, it shouldn't be considered optional.
+      @JsonProperty("visibility") @Nullable Segments visibility
+  )
+  {
+    this.dataSource = dataSource;
+    this.interval = interval;
+    // Defaulting to the former behaviour when visibility wasn't explicitly specified for backward compatibility
+    this.visibility = visibility != null ? visibility : Segments.ONLY_VISIBLE;
+  }
+
+  @JsonProperty
+  public String getDataSource()
+  {
+    return dataSource;
+  }
+
+  @JsonProperty
+  public Interval getInterval()
+  {
+    return interval;
+  }
+
+  @JsonProperty
+  public Segments getVisibility()
+  {
+    return visibility;
+  }
+
+  @Override
+  public TypeReference<Collection<DataSegment>> getReturnTypeReference()
+  {
+    return new TypeReference<Collection<DataSegment>>() {};
+  }
+
+  @Override
+  public Collection<DataSegment> perform(Task task, TaskActionToolbox toolbox)
+  {
+    final Set<ReplaceTaskLock> replaceLocksForTask = toolbox.getTaskLockbox().findReplaceLocksForTask(task);
+    String createdBefore = null;
+    if (task.getDataSource().equals(dataSource)) {
```

Comment:
This should be a precondition check at the start of this method.

Reply:
A reindexing job utilizing the action could read data from one datasource, but write to another

Reply:
Makes sense 👍🏻